src/pools.jl

```diff
@@ -0,0 +1,109 @@
+`release(pool)` will return the "permit" to the pool while not returning
+any object for reuse.
+"""
+struct Pool{K, T}
+    sem::Semaphore
+    values::Lockable{Dict{K, ConcurrentStack{T}}}
+end
+
+Pool(T; max::Int=typemax(Int)) = Pool{Nothing, T}(Semaphore(max), Lockable(Dict{Nothing, ConcurrentStack{T}}()))
+Pool(K, T; max::Int=typemax(Int)) = Pool{K, T}(Semaphore(max), Lockable(Dict{K, ConcurrentStack{T}}()))
+
+Base.empty!(pool::Pool) = Base.@lock pool.values empty!(pool.values[])
+
+TRUE(x) = true
+
+"""
+    acquire(f, pool::Pool{K, T}, [key::K]; forcenew::Bool=false, isvalid::Function) -> T
+
+Get an object from a `pool`, optionally keyed by the provided `key`. If no pool exists for the given key, one will be created.
+The provided function `f` must create a new object instance of type `T`.
+The acquired object MUST be returned to the pool by calling `release(pool, key, obj)` exactly once.
+The `forcenew` keyword argument can be used to force the creation of a new object, ignoring any existing objects in the pool.
+The `isvalid` keyword argument can be used to specify a function that will be called to determine if an object is still valid
+for reuse. By default, all objects are considered valid.
+If there are no objects available for reuse, `f` will be called to create a new object.
+If the pool is already at its maximum capacity, `acquire` will block until an object is returned to the pool via `release`.
+"""
+function Base.acquire(f, pool::Pool{K, T}, key=nothing; forcenew::Bool=false, isvalid::Function=TRUE) where {K, T}
+    key isa K || throw(ArgumentError("invalid key `$key` provided for pool key type $K"))
+    acquire(pool.sem)
+    try
+        # once we've received our permit, we can figure out where the object should come from
+        # if we're forcing a new object, we can just call f() and return it
+        forcenew && return f()
+        # otherwise, check if there's an existing object in the pool to reuse
+        objs = Base.@lock pool.values get!(() -> ConcurrentStack{T}(), pool.values[], key)
+        obj = pop!(objs)
+        while obj !== nothing
+            # if the object is valid, return it
+            isvalid(obj) && return obj
+            # otherwise, try the next object
+            obj = pop!(objs)
+        end
+        # if we get here, we didn't find any valid objects, so we'll just create a new one
+        return f()
+    catch
+        # an error occurred while acquiring, so make sure we return the semaphore permit
+        release(pool.sem)
+        rethrow()
+    end
+end
+
+"""
+    release(pool::Pool{K, T}, key::K, obj::Union{T, Nothing}=nothing)
+    release(pool::Pool{K, T}, obj::T)
+    release(pool::Pool{K, T})
+
+Return an object to a `pool`, optionally keyed by the provided `key`.
+If `obj` is provided, it will be returned to the pool for reuse. Otherwise, if `nothing` is returned,
+just the "permit" will be returned to the pool.
```

Comment:
so for a keyed pool to return just the permit you have to call `release(pool, key, nothing)` but for a non-keyed pool you have to call `release(pool)`? that seems inconsistent to me

Reply:
I can update the comment, but you can call `release(pool, key)` on a keyed pool.